management/email_provider.go

```diff
@@ -0,0 +1,288 @@
+}
+
+// EmailProviderCredentialsMailgun represent the
+// credentials required to use the mailgun provider.
+type EmailProviderCredentialsMailgun struct {
+	APIKey *string `json:"api_key,omitempty"`
+	Domain *string `json:"domain,omitempty"`
+	Region *string `json:"region,omitempty"`
+}
+
+// EmailProviderCredentialsSMTP represent the
+// credentials required to use the smtp provider.
+type EmailProviderCredentialsSMTP struct {
+	SMTPHost *string `json:"smtp_host,omitempty"`
+	SMTPPort *int    `json:"smtp_port,omitempty"`
+	SMTPUser *string `json:"smtp_user,omitempty"`
+	SMTPPass *string `json:"smtp_pass,omitempty"`
+}
+
+// EmailProviderSettingsMandrill are the provider
+// specific settings used by the mandrill provider.
+type EmailProviderSettingsMandrill struct {
+	Message *EmailProviderSettingsMandrillMessage `json:"message,omitempty"`
+}
+
+// EmailProviderSettingsMandrillMessage contains the
+// message settings content for the mandrill provider.
+type EmailProviderSettingsMandrillMessage struct {
+	ViewContentLink *bool `json:"view_content_link,omitempty"`
+}
+
+// EmailProviderSettingsSES are the provider
+// specific settings used by the ses provider.
+type EmailProviderSettingsSES struct {
+	Message *EmailProviderSettingsSESMessage `json:"message,omitempty"`
+}
+
+// EmailProviderSettingsSESMessage contains the
+// message settings content for the ses provider.
+type EmailProviderSettingsSESMessage struct {
+	ConfigurationSetName *string `json:"configuration_set_name,omitempty"`
+}
+
+// EmailProviderSettingsSMTP are the provider
+// specific settings used by the smtp provider.
+type EmailProviderSettingsSMTP struct {
+	Headers *EmailProviderSettingsSMTPHeaders `json:"headers,omitempty"`
+}
+
+// EmailProviderSettingsSMTPHeaders contains the
+// headers settings content for the smtp provider.
+type EmailProviderSettingsSMTPHeaders struct {
+	XMCViewContentLink   *string `json:"X-MC-ViewContentLink,omitempty"`
+	XSESConfigurationSet *string `json:"X-SES-Configuration-Set,omitempty"`
+}
+
+// MarshalJSON is a custom serializer for the EmailProvider type.
+func (ep *EmailProvider) MarshalJSON() ([]byte, error) {
+	type emailProvider EmailProvider
+	type emailProviderWrapper struct {
+		*emailProvider
+		RawCredentials json.RawMessage `json:"credentials,omitempty"`
+		RawSettings    json.RawMessage `json:"settings,omitempty"`
+	}
+
+	wrapper := &emailProviderWrapper{(*emailProvider)(ep), nil, nil}
+
+	if ep.Credentials != nil {
+		credentialsJSON, err := json.Marshal(ep.Credentials)
+		if err != nil {
+			return nil, err
+		}
+		wrapper.RawCredentials = credentialsJSON
+	}
+
+	if ep.Settings != nil {
+		settingsJSON, err := json.Marshal(ep.Settings)
+		if err != nil {
+			return nil, err
+		}
+		wrapper.RawSettings = settingsJSON
+	}
+
+	return json.Marshal(wrapper)
+}
+
+// UnmarshalJSON is a custom deserializer for the EmailProvider type.
+func (ep *EmailProvider) UnmarshalJSON(b []byte) error {
+	type emailProvider EmailProvider
```

Comment:
Maybe move this outside to avoid duplicating it? It's also in `MarshalJSON()`.

Reply:
Fair point, this is now available in [`f9ace1c` (#129)](https://github.com/auth0/go-auth0/pull/129/commits/f9ace1ceb6d7007c51572abb73ac576d52cf02f4)